### ISSUE TITLE
Question: why doesn't this work?

### ISSUE THREAD

**Issue by kashav** [microsoft/pyright#10273]
```python3
a = []
a.append(1)
reveal_type(a)

b = [1]
reveal_type(b)
```

```terminal
% ./env/bin/pyright a.py
a.py
  a.py:33:13 - information: Type of "a" is "list[Unknown]"
  a.py:36:13 - information: Type of "b" is "list[int]"
0 errors, 0 warnings, 2 informations
% ./env/bin/mypy a.py
a.py:33: note: Revealed type is "builtins.list[builtins.int]"
a.py:36: note: Revealed type is "builtins.list[builtins.int]"
Success: no issues found in 1 source file
```

mypy reports `list[int]` but pyright reports `list[Unknown]`. Why doesn't this work? Is this easy to fix?

**Comment by kashav** [comment]
FWIW and if i'm not mistaken, for the following code snippet the latter `foo.foo` reference doesn't link to the getter in VSCode because of how this typechecks in pyright:

```python3
from typing import List

class Foo:
    def __init__(self):
        self._foo = 1

    @property
    def foo(self) -> int:
        return self._foo

foo = Foo()
x = foo.foo

Foos: List[Foo] = []
for _ in range(10):
    Foos.append(Foo())

for foo in Foos:
    x = foo.foo

def get_foo() -> Foo:
    return Foo()

Foos2 = []
for _ in range(10):
    Foos2.append(get_foo())

for foo in Foos2:
    x = foo.foo # this doesn't link to the getter function in VSCode
```

**Comment by erictraut** [comment]
The type of `[]` is inferred as `list[Unknown]` because there is insufficient type information to infer something more precise. Pyright assumes that the type of an object must be established at the time that it's constructed. Its specialized type is then used to subsequently evaluate the specialized type of methods and attributes on the object. For example, if the type of `a` is determined to be `list[int]` at the time of construction, then the type of the expression `a.append` evaluates to `(object: int) -> None`, and it's legal to call with an `int` argument (but illegal to call it with a `str` argument, for example). The type of `a` cannot change based on the method that you call on it; that would be backward.

I'm guessing that mypy breaks this rule and hard-codes some knowledge of the `list.append` method under certain specific circumstances. If so, that's pretty hacky — not something I'd be inclined to do in pyright. You can see how it falls apart quickly. For example, if insert an extra `reveal_type(x)` call (which shouldn't affect type evaluation), mypy's behavior changes.

```python
x = [] # Error: Need type annotation for "x"
reveal_type(x)
x.append(1)
reveal_type(x)
```

Another difference between mypy and pyright is that pyright doesn't use top-to-bottom type evaluation like mypy. It performs a just-in-time type evaluation for any expression in your code. If you'd like to better understand the reasons for this design choice, refer to [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=design-goals). An implication of this design decision is that the evaluated type of an expression can be affected by code that is "upstream" of an expression within the code flow graph but not "downstream". For example, a [type guard](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards) can affect the subsequent type of an expression through type narrowing. However, an expression's type can't be changed by code that is "downstream" within the code flow graph. In your first example, the type of `a` in the statement `a = []` can't be affected by the subsequent statement `a.append(1)`.

I suppose it would be theoretically possible to support this in a just-in-time evaluator, but it would be very expensive from a memory and computation perspective. It would require building twice as many code flow graphs (to cover both directions) and performing code flow analysis in both directions for every expression's type evaluation. That would be very costly.

Here are some ways to write your code in a way that establishes the type at construction time.

```python
# Declare type for assignment target
x: list[int] = []

# Call a specialized constructor
x = list[int]()

# Provide an element in the list expression rather than appending afterward
x = [1]
```
